libs/galley-types/src/Galley/Types/Teams.hs

```diff
@@ -0,0 +1,472 @@
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StrictData        #-}
+{-# LANGUAGE TemplateHaskell   #-}
+
+module Galley.Types.Teams
+    ( Team
+    , newTeam
+    , teamId
+    , teamCreator
+    , teamName
+    , teamIcon
+    , teamIconKey
+
+    , TeamList
+    , newTeamList
+    , teamListTeams
+    , teamListHasMore
+
+    , TeamMember
+    , newTeamMember
+    , userId
+    , permissions
+    , teamMemberJson
+
+    , TeamMemberList
+    , newTeamMemberList
+    , teamMembers
+    , teamMemberListJson
+
+    , TeamConversation
+    , newTeamConversation
+    , conversationId
+    , managedConversation
+
+    , TeamConversationList
+    , newTeamConversationList
+    , teamConversations
+
+    , Permissions
+    , newPermissions
+    , fullPermissions
+    , hasPermission
+    , self
+    , copy
+
+    , Perm (..)
+    , permToInt
+    , permsToInt
+    , intToPerm
+    , intToPerms
+
+    , NewTeam
+    , newNewTeam
+    , newTeamName
+    , newTeamIcon
+    , newTeamIconKey
+    , newTeamMembers
+
+    , NewTeamMember
+    , newNewTeamMember
+    , ntmNewTeamMember
+
+    , Event
+    , newEvent
+    , eventType
+    , eventTime
+    , eventTeam
+    , eventData
+
+    , EventType (..)
+    , EventData (..)
+
+    , TeamUpdateData
+    , newTeamUpdateData
+    , nameUpdate
+    , iconUpdate
+    , iconKeyUpdate
+
+    ) where
+
+import Control.Lens (makeLenses, (^.))
+import Control.Monad (when)
+import Data.Aeson
+import Data.Aeson.Types (Parser)
+import Data.Bits (testBit, (.|.))
+import Data.Id (TeamId, ConvId, UserId)
+import Data.Json.Util
+import Data.Monoid
+import Data.Maybe (mapMaybe, isNothing)
+import Data.Range
+import Data.Set (Set)
+import Data.Text (Text)
+import Data.Time (UTCTime)
+import Data.Word
+
+
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.Set as Set
+
+data Team = Team
+    { _teamId      :: TeamId
+    , _teamCreator :: UserId
+    , _teamName    :: Text
+    , _teamIcon    :: Text
+    , _teamIconKey :: Maybe Text
```

Comment:
`teamIconKey` denotes the asset key in case the team icon asset is not meant to be public.